### PR TITLE
[MRG] ENH: modify time_as_index to allow arbitrary origin

### LINKED ISSUE #5523: time_as_index
Now that `raw.annotations.orig_time` is in sync with `raw.info['meas_date']`, the `raw.time_as_index` might be confusing. See the following:

```py
	from mne.io import RawArray
	from mne import create_info, Annotations

    meas_date=1
    info = create_info(ch_names=10, sfreq=10.)
    raw = RawArray(data=np.empty((10, 10)), info=info, first_samp=10)
    raw.info['meas_date'] = meas_date

    annot = Annotations(onset=raw.times,
                        duration=np.zeros_like(raw.times),
                        description='dummy',
                        orig_time=None)

    raw.set_annotations(annotations=annot)

    print(raw.times)
    print(raw.annotations.onset)
    print(raw.time_as_index(raw.times))
    print(raw.time_as_index(raw.annotations.onset))
```

```sh
[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9]
[1.  1.1 1.2 1.3 1.4 1.5 1.6 1.7 1.8 1.9]
[0 1 2 3 4 5 6 7 8 9]
[10 11 12 13 14 15 16 17 18 19]
```
Shall we add a reference parameter to `time_as_index`?

### PR DESCRIPTION
**Pull request by massich**
Fixes #5523 

This PR is required for #4936

### CODE REVIEW

File: mne/io/tests/test_raw.py [hunk 1]
```diff
@@ -151,7 +151,7 @@ def test_time_index():
 
     # Test new (rounding) indexing behavior
     new_inds = raw.time_as_index(raw.times, use_rounding=True)
-    assert(len(set(new_inds)) == len(new_inds))
+    assert_array_equal(new_inds, np.array(range(len(raw.times))))
```
Comment (agramfort) [inline mne/io/tests/test_raw.py]: np.arange

**Comment by massich** [comment]
Now it should be green. @agramfort 

The fact that [this](https://travis-ci.org/mne-tools/mne-python/jobs/428665877#L2843) failed might suggest that passing `origin=xxxx` to `my_epochs.tima_as_index` might produce undesired results. Looking at the blame, @dengemann and @larsoner might have some saying in adding this feature.

**Comment by larsoner** [comment]
I don't think `epochs.time_as_index` should have an `origin` arg. It doesn't make sense there, or in `Evoked`, `SourceEstimate`, or any other object -- except for `BaseRaw`. So you should probably override the method in `BaseRaw` to call the `super` method, and then adjust the time indices to respect the `origin` arg.

**Comment by agramfort** [comment]
@larsoner ok for you?

File: mne/io/base.py [hunk 2]
```diff
@@ -650,6 +650,40 @@ def last_samp(self):
...
+        if origin is None:
+            origin = first_samp_in_abs_time
+
+        absolute_time = np.atleast_1d(times) + _handle_meas_date(origin)
+        index = (absolute_time - first_samp_in_abs_time) * self.info['sfreq']
+
+        if use_rounding:
+            index = np.round(index)
+
+        return index.astype(int)
```
Comment (larsoner) [inline mne/io/base.py]: It seems more DRY to convert whatever time + origin input to be in "time relative to first sample", then call the super method:
```
... 
time_rel_first_samp = ...
return super(BaseRaw, self).time_as_index(time_rel_first
_samp, use_rounding)
```
It gets rid of redundancies in multiplying by sfreq, triaging use_rounding, and output dtype.

**Comment by agramfort** [comment]
thx @massich

**Pull request merged**